src/Compiler/Service/TransparentCompiler.fs

```diff
@@ -0,0 +1,987 @@
+
+            let diagnosticsLogger =
+                CompilationDiagnosticLogger("Parse", tcConfig.diagnosticsOptions)
+            // Return the disposable object that cleans up
+            use _holder = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.Parse)
+
+            let flags = file.IsLastCompiland, file.IsExe
+            let fileName = file.Source.FileName
+            let! sourceText = file.Source.GetSource() |> NodeCode.AwaitTask
+
+            let input =
+                ParseOneInputSourceText(tcConfig, lexResourceManager, fileName, flags, diagnosticsLogger, sourceText)
+
+            return input, diagnosticsLogger.GetDiagnostics(), sourceText
+        }
+
+    let ComputeDependencyGraphForLastFile parsedInputs (tcConfig: TcConfig) _key =
+        node {
+            let sourceFiles: FileInProject array =
+                parsedInputs
+                |> Seq.toArray
+                |> Array.mapi (fun idx (input: ParsedInput) ->
+                    {
+                        Idx = idx
+                        FileName = input.FileName
+                        ParsedInput = input
+                    })
+
+            let filePairs = FilePairMap(sourceFiles)
+
+            // TODO: we will probably want to cache and re-use larger graphs if available
+            let graph =
+                DependencyResolution.mkGraph tcConfig.compilingFSharpCore filePairs sourceFiles
+                |> fst
+                |> Graph.subGraphFor (sourceFiles |> Array.last).Idx
+
+            return graph, filePairs
+        }
+
+    let ComputeTcIntermediate (parsedInput: ParsedInput, parseErrors) bootstrapInfo prevTcInfo _key =
+        node {
+            let input = parsedInput
+            let fileName = input.FileName
+            let tcConfig = bootstrapInfo.TcConfig
+            let tcGlobals = bootstrapInfo.TcGlobals
+            let tcImports = bootstrapInfo.TcImports
+
+            let capturingDiagnosticsLogger = CapturingDiagnosticsLogger("TypeCheck")
+
+            let diagnosticsLogger =
+                GetDiagnosticsLoggerFilteringByScopedPragmas(
+                    false,
+                    input.ScopedPragmas,
+                    tcConfig.diagnosticsOptions,
+                    capturingDiagnosticsLogger
+                )
+
+            use _ = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.TypeCheck)
+
+            //beforeFileChecked.Trigger fileName
+
+            ApplyMetaCommandsFromInputToTcConfig(tcConfig, input, Path.GetDirectoryName fileName, tcImports.DependencyProvider)
+            |> ignore
+
+            let sink = TcResultsSinkImpl(tcGlobals)
+            let hadParseErrors = not (Array.isEmpty parseErrors)
+
+            let input, moduleNamesDict =
+                DeduplicateParsedInputModuleName prevTcInfo.moduleNamesDict input
+
+            let! (tcEnvAtEndOfFile, topAttribs, implFile, ccuSigForFile), tcState =
+                CheckOneInput(
```

Comment:
Probably `CheckOneInputWithCallback` will be the way to go. With some care to not get duplicated stuff in prior tcState.

Reply:
I would prefer to solve it without callbacks for increased simplicity. But if it doesn't work out then I will probably use that one.